src/s3/async_put_object.rs

```diff
@@ -106,12 +106,8 @@ impl<'a> AsyncWrite for AsyncPutObject<'a> {
                     .body(mem::take(&mut self.buf).into())
                     .acl(ObjectCannedAcl::BucketOwnerFullControl)
                     .send();
-                self.state = PutObjectState::Closing(Box::pin(fut));
-                cx.waker().wake_by_ref();
-                Poll::Pending
-            }
-            PutObjectState::Closing(ref mut fut) => {
-                match Pin::new(fut).poll(cx) {
+                let mut fut = Box::pin(fut);
```

Comment:
Github won't let me do this as a suggestion, but if we remove the wake at line 124, and take advantage of the `ready!` macro, we can boil the whole thing down to:

```
                self.state = PutObjectState::Closing(Box::pin(fut));
                // Manually trigger a wake so that the executor immedially
                // polls us again, which will take us into the `Closing` block.
                cx.waker().wake_by_ref();
                Poll::Pending
            }
            PutObjectState::Closing(ref mut fut) => {
                let result = ready!(Pin::new(fut).poll(cx));
                self.state = PutObjectState::Closed;
                Poll::Ready(match result {
                    Ok(_) => Ok(()),
                    Err(e) => Err(Error::new(ErrorKind::Other, e)),
                })
            }
```

Reply:
Ah, very nice, will update the PR to match 👍🏻